### PR TITLE
[DROOLS-6821] Inconsistent propety reactivity with Join and From

### PR DESCRIPTION
**Pull request by tkobayas**
- Unit test only

With the current implementation, `testJoinAndFrom` fails and `testFromAndJoin` passes. 

**Ports** 
This PR is for 7.x.
Will do for main.

**JIRA**: 
https://issues.redhat.com/browse/DROOLS-6821

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* <b>a pull request</b> please add comment: <b>Jenkins retest this</b>

* for a <b>full downstream build</b> 
  * for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  * for <b>github actions</b> job: add the label `run_fdb`

* <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

* <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

* <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>
</details>

### CODE REVIEW

**Comment by tkobayas** [comment]
@mariofusco Please review.

See `FromTest.testJoinAndFrom()`. In R2, the 1st Person is simply passed to the 2nd Person (Odd but able to write). In this case, the 1st Person's LeftInputAdapterNode's sinkMask is empty.
```
rule R1
    ...
    when
        $p1 : Person ( )
        $s : String(this == "ABC")
...

rule R2
    ...
    when
        $p1 : Person ( )
        $p2 : Person ( age < 2 ) from $p1

```

Then the LeftInputAdapterNode is eventually shared for R1's JoinNode and R2's FromNode.
```
  [ObjectTypeNode(3)::EntryPoint::DEFAULT objectType=[ClassObjectType class=org.drools.testcoverage.common.model.Person] expiration=-1ms ] on Partition(MAIN)
    [AlphaNode(11) constraint=[Constraint for 'GENERATED_EB809CE05631DA5BE6C15A7A3366B49C' (index: AlphaIndex #2 (EQUAL, left: lambda 269973396, right: 1)), age == 1]] on Partition(4) d 8 i 8
      [AlphaTerminalNode(12)] on Partition(4) Ld 0 Li 0
        [RuleTerminalNode(13): rule=R3] on Partition(4) d 0 i 8
    [LeftInputAdapterNode(4)] on Partition(3) Ld 0 Li 0
      [JoinNode(7) - [ClassObjectType class=java.lang.String]] <contraints=[]>  on Partition(3) Ld 0 Li 0 Rd -1 Ri -1
        [RuleTerminalNode(8): rule=R1] on Partition(3) d -1 i -1
      [FromNode(9)] <result=org.drools.testcoverage.common.model.Person, alphaConstraints=[[Constraint for 'GENERATED_0B8E6416B245FDA0739C49A931B0F323' (index: AlphaIndex #2 (LESS_THAN, left: lambda 269973396, right: 2)), age < 2]], betaConstraints=[]>  on Partition(3) Ld 8 Li 8
        [RuleTerminalNode(10): rule=R2] on Partition(3) d 0 i 0
```
In this case, property reactivity depends on the order of JoinNode (mask = 0) and FromNode (mask = 8), which have different masks because `LeftInputAdapterNode.modifyObject()` relies on the first sink's mask.

https://github.com/kiegroup/drools/blob/7.65.0.Final/drools-core/src/main/java/org/drools/core/reteoo/LeftInputAdapterNode.java#L411-L422

I think the right way to fix is to have different LeftInputAdapterNodes for the JoinNode and the FromNode. So I enhanced `RuleDescr.lookAheadFieldsOfIdentifier()` for the special case where `from` is used to directly pass a Pattern to the next Pattern.

With the change, the property (`age`) is added to `listenedProperties` of the 1st Pattern so that a `LeftInputAdapterNode` will have a correct `sinkMask`. Then, the `LeftInputAdapterNode` is not shared with the R1's JoinNode.

```
  [ObjectTypeNode(3)::EntryPoint::DEFAULT objectType=[ClassObjectType class=org.drools.testcoverage.common.model.Person] expiration=-1ms ] on Partition(MAIN)
    [AlphaNode(12) constraint=[Constraint for 'GENERATED_EB809CE05631DA5BE6C15A7A3366B49C' (index: AlphaIndex #2 (EQUAL, left: lambda 1096433514, right: 1)), age == 1]] on Partition(6) d 8 i 8
      [AlphaTerminalNode(13)] on Partition(6) Ld 0 Li 0
        [RuleTerminalNode(14): rule=R3] on Partition(6) d 0 i 8
    [LeftInputAdapterNode(4)] on Partition(4) Ld 0 Li 0
      [JoinNode(7) - [ClassObjectType class=java.lang.String]] <contraints=[]>  on Partition(4) Ld 0 Li 0 Rd -1 Ri -1
        [RuleTerminalNode(8): rule=R1] on Partition(4) d -1 i -1
    [LeftInputAdapterNode(9)] on Partition(5) Ld 0 Li 0
      [FromNode(10)] <result=org.drools.testcoverage.common.model.Person, alphaConstraints=[[Constraint for 'GENERATED_0B8E6416B245FDA0739C49A931B0F323' (index: AlphaIndex #2 (LESS_THAN, left: lambda 1096433514, right: 2)), age < 2]], betaConstraints=[]>  on Partition(5) Ld 8 Li 8
        [RuleTerminalNode(11): rule=R2] on Partition(5) d 0 i 0
```

File: drools-compiler/src/main/java/org/drools/compiler/lang/descr/RuleDescr.java [hunk 1]
```diff
@@ -371,6 +371,37 @@ public boolean equals(Object obj) {
...
+
+    private static Collection<String> collectProps(PatternDescr pattern) {
+        Collection<String> props = new HashSet<>();
+        for (BaseDescr expr : pattern.getDescrs()) {
+            if (expr instanceof ExprConstraintDescr) {
+                String text = expr.getText();
+                String prop = extractFirstIdentifier(text, 0);
+                String propFromGetter = getter2property(prop);
+                props.add(propFromGetter != null ? propFromGetter : lcFirst(prop));
             }
```
Comment (tkobayas) [inline drools-compiler/src/main/java/org/drools/compiler/lang/descr/RuleDescr.java:405]: This text parsing can cover only simple field constraints. I may be able to write more complex test cases to harden this method but not sure if we want to work a lot on such an odd use case...

**Comment by tkobayas** [comment]
Linux - Pull Request : Failed test is not related to this PR.
```
org.kie.dmn.feel.client.showcase.FEELShowcaseIT.org.kie.dmn.feel.client.showcase.FEELShowcaseIT
```

**Comment by mariofusco** [comment]
@tkobayas I saw what you did, and yes I wouldn't do more than that for such an odd use case. In all honesty if things become more complicated than this I'd rather raise a compile time error because indeed this use of `from` is really meaningless. For now let's keep it this way. When you'll forward port this fix to main branch be aware that I moved somewhere else that `lookAheadFieldsOfIdentifier` method when I isolated the ast in its own module with this commit https://github.com/kiegroup/drools/commit/b7ba8c1b34d312a578150876b7c049af93b799a4

**Review by mariofusco — APPROVED** [review]

**Pull request merged**